src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -634,84 +634,166 @@ private static object SplitOperatorImpl(ExecutionContext context, IScriptExtent
+                    // Evaluate the predicate using the character at cursor.
+                    object predicateResult = predicate.DoInvokeReturnAsIs(
+                        useLocalScope: true,
+                        errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToExternalErrorPipe,
+                        dollarUnder: CharToString(item[cursor]),
+                        input: AutomationNull.Value,
+                        scriptThis: AutomationNull.Value,
+                        args: new object[] { item, cursor });
+
+                    if (!LanguagePrimitives.IsTrue(predicateResult))
+                    {
+                        subStringLength++;
+                        cursor -= 1;
+                        continue;
+                    }
+
+                    split.Add(item.Substring(cursor + 1, subStringLength));
+
+                    subStringLength = 0;
 
-                if (limit == 1)
+                    cursor -= 1;
+
+                    if (System.Math.Abs(limit) == (split.Count + 1))
+                    {
+                        break;
+                    }
+                }
+
+                if (cursor == -1)
                 {
-                    // Don't bother with looking for any delimiters,
-                    // just return the original string.
-                    results.Add(item);
-                    continue;
+                    // Used when the limit is negative
+                    // and the cursor was allowed to go
+                    // all the way to the start of the
+                    // string.
+                    split.Add(item.Substring(0, subStringLength));
                 }
+                else
+                {
+                    // Used to get the rest of the string
+                    // when using a negative limit and
+                    // the cursor doesn't reach the end
+                    // of the string.
+                    split.Add(item.Substring(0, cursor + 1));
+                }
+
+                split.Reverse();
+
+                results.AddRange(split);
+            }
+
+            return results.ToArray();
+        }
+
+        private static IReadOnlyList<string> SplitWithPredicate(ExecutionContext context, IScriptExtent errorPosition, IEnumerable<string> content, ScriptBlock predicate, int limit)
+        {
+            var results = new List<string>();
+
+            if (limit == 1)
+            {
+                // If the user just wants 1 string
+                // then just return the content
+                return new List<string>(content);
+            }
+
+            foreach (string item in content)
+            {
+                var split = new List<string>();
 
-                StringBuilder buf = new StringBuilder();
-                for (int strIndex = 0; strIndex < item.Length; strIndex++)
+                // Used to traverse through the item
+                var cursor = 0;
```

Comment:
```suggestion
                int cursor = 0;
```

Reply:
why? because it's a primitive?

Reply:
Just because using `var` doesn't save any characters, I think, and it doesn't hurt to be explicit here.

Reply:
Makes sense to me I just missed a couple on the previous commit but I fixed it with the most recent one.

Reply:
Well `var` doesn't really save characters and unlike with a structured type, there's no substitution principle going on (where the `var` is a way of saying "I don't care what the type is as long as it has this method or can be passed through this interface").

Also, primitive literals don't always have obvious types (in C#, the type of an integer literal without a suffix [is the smallest integer type that will contain it](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/int)). (In C, `'a'` is of type `int`...)